core/src/main/scala/quasar/plugin/postgres/CsvSink.scala

```diff
@@ -0,0 +1,213 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package quasar.plugin.postgres
+
+import slamdata.Predef.{Stream => _, _}
+
+import cats._
+import cats.data._
+import cats.effect.{Effect, ExitCase, LiftIO, Sync, Timer}
+import cats.effect.concurrent.Ref
+import cats.implicits._
+
+import doobie._
+import doobie.implicits._
+import doobie.postgres._
+import doobie.postgres.implicits._
+import doobie.util.log.{ExecFailure, ProcessingFailure, Success}
+
+import fs2.{Chunk, Pipe, Stream}
+
+import java.lang.IllegalArgumentException
+
+import org.slf4s.Logging
+
+import quasar.connector._
+import quasar.api.resource._
+import quasar.api.table.{ColumnType, TableColumn}
+
+import scala.concurrent.duration.MILLISECONDS
+
+import shims._
+
+object CsvSink extends Logging {
+  def apply[F[_]: Effect: MonadResourceErr](
+      xa: Transactor[F],
+      writeMode: WriteMode)(
+      dst: ResourcePath,
+      columns: List[TableColumn],
+      data: Stream[F, Byte])(
+      implicit timer: Timer[F])
+      : F[Unit] = {
+
+    val AE = ApplicativeError[F, Throwable]
+
+    val table =
+      tableFromPath(dst) match {
+        case Some(t) =>
+          t.pure[F]
+
+        case None =>
+          MonadResourceErr[F].raiseError(ResourceError.notAResource(dst))
+      }
+
+    val tableColumns =
+      columns.toNel match {
+        case Some(cols) =>
+          cols.pure[F]
+
+        case None =>
+          AE.raiseError(new IllegalArgumentException("No columns specified."))
+      }
+
+    for {
+      tbl <- table
+
+      cols <- tableColumns
+
+      action = writeMode match {
+        case WriteMode.Create => "Creating"
+        case WriteMode.Replace => "Replacing"
+      }
+
+      _ <- debug[F](s"${action} '${tbl}' with schema ${cols.show}")
+
+      // Telemetry
+      totalBytes <- Ref[F].of(0L)
+      startAt <- timer.clock.monotonic(MILLISECONDS)
+
+      doCopy =
+        data
+          .chunks
+          .evalTap(recordChunks[F](totalBytes))
+          // TODO: Is there a better way?
+          .translate(Effect.toIOK[F] andThen LiftIO.liftK[CopyManagerIO])
+          .through(copyToTable(tbl, cols))
+          .compile
+          .drain
+
+      ensureTable =
+        writeMode match {
+          case WriteMode.Create =>
+            createTable(tbl, cols)
+
+          case WriteMode.Replace =>
+            dropTableIfExists(tbl) >> createTable(tbl, cols)
+        }
+
+      _ <- (ensureTable >> PHC.pgGetCopyAPI(doCopy)).transact(xa) handleErrorWith { t =>
+        error[F](s"COPY to '${tbl}' produced unexpected error: ${t.getMessage}", t) >>
+          AE.raiseError(t)
```

Comment:
Style note: is there any reason to use `>>` inside of a for comprehension like this? Why not just have three `_ <-` lines?

Reply:
The `>>` is within the `handleErrorWith` block, it isn't part of the for-comprehension.

Reply:
Ah